sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DecorrelateInnerQuery.scala

```diff
@@ -667,6 +720,24 @@ object DecorrelateInnerQuery extends PredicateHelper {
             val newJoin = j.copy(left = newLeft, right = newRight, condition = newCondition)
             (newJoin, newJoinCond, newOuterReferenceMap)
 
+          case u: Union =>
+            // First collect outer references from all children
+            val collectedChildOuterReferences = collectOuterReferencesInPlanTree(u)
```

Comment:
Why do we need to collect all outer references from all children first? Can you add some high-level ideas on how decorrelate through Union works? Preferable with examples.

Reply:
It's described with an example in the linked doc https://docs.google.com/document/d/11b9ClCF2jYGU7vU2suOT7LRswYkg6tZ8_6xJbvxfh2I/edit# - does that answer your question? I'll add the info to the comments as well.